type-inference-in-readr-and-arrow.rmd

```diff
@@ -0,0 +1,271 @@
+parse an input as that data type. For example, for a column to be of
+logical type, it can only contain `TRUE`, `FALSE`, or `NA` values, which
+is why this is the most strict type, but all of the other types could be
+read in as character data, which is the least strict and why this is
+last in the order.
+
+## How does type inference work in arrow?
+
+In arrow, `read_csv_arrow()` handles CSV reading, and much of its
+interface has been designed to closely follow the excellent APIs of
+`readr::read_csv()` and `vroom::vroom()`. The intention here is that
+users can use parameter names they're familiar with from the
+aforementioned readers when using arrow, and get the same results. The
+underlying code is pretty different though, and type inference is done
+on the whole column rather than an initial subset.
+
+In addition, Arrow has a different set of possible data types compared
+to R; see [the Arrow docs](https://arrow.apache.org/docs/r/articles/arrow.html) for more
+information about the mapping between R data types and Arrow types.
+
+In [the Arrow docs](https://arrow.apache.org/docs/cpp/csv.html#data-types), we can see
+that types are inferred in this order:
+
+-   Null
+
+-   Int64
+
+-   Boolean
+
+-   Date32
+
+-   Timestamp (with seconds unit)
+
+-   Float64
+
+-   Dictionary\<String\> (if ConvertOptions::auto_dict_encode is true)
+
+-   Dictionary\<Binary\> (if ConvertOptions::auto_dict_encode is true)
+
+-   String
+
+-   Binary
+
+Note that if you use `arrow::read_csv_arrow()` with parameter
+`as_data_frame = TRUE` (the default), the Arrow data types are then
+converted to R data types.
+
+```{r}
+simple_data <- data.frame(x = c(1, 2, 3), y = c("a", "b", "c"), z = c(1.1, 2.2, 3.3))
+
+readr::write_csv(simple_data, "simple_data.csv")
+
+# columns are arrow's int64, string, and double (aka float64) types
+arrow::read_csv_arrow("simple_data.csv", as_data_frame = FALSE)
+```
+
+```{r}
+# columns converted to R's integer, character, and double types
+arrow::read_csv_arrow("simple_data.csv", as_data_frame = TRUE)
+```
+
+## What are the main differences between readr and arrow type inference?
+
+Although there appear to be quite a few differences between the order of type inference when comparing arrow and readr, in practice, this doesn’t have much effect.  Type inference for logical/boolean and integer values are the opposite way round, but given that the underlying data that translates into these types looks very different, they’re not going to be mixed up.  The biggest differences come from custom behaviours which are specific to readr and arrow; I’ve outlined them below.
+
+### Guessing integers
+
+In the code for readr, the default setting is for numeric values to always be read in as doubles but never integers.  If you want readr to guess that a column may be an integer, you need to read it in as character data, and then call `type_convert()`.  In arrow, if data can be represented as integers but not doubles, then it will be.
```

Comment:
I'm hesitant to swap it out directly, as this is all about inference, but I agree that the `type_convert()` thing is weird, so have added a bit in that basically says "just specify the column type".

Reply:
Oh, I see you are really trying to show *guessing* specifically.

Sidebar: It's a very conscious choice in vroom/readr (and it wasn't always this way!) to **not** ever guess an integer type. There were too many instances of people building a pipeline with a toy dataset and having some column discovered to be integer. Then "in production" there would be new data and more data and it turned out the column doesn't necessarily contain only integers, so it should really be double. And then that would cause trouble. Hence, integer types can be specified but are never guessed.